middleware/markdown/process.go

```diff
@@ -1,216 +1,72 @@
 package markdown
 
 import (
-	"bytes"
 	"io/ioutil"
-	"log"
 	"os"
 	"path/filepath"
-	"strings"
-	"text/template"
 
 	"github.com/mholt/caddy/middleware"
+	"github.com/mholt/caddy/middleware/markdown/metadata"
+	"github.com/mholt/caddy/middleware/markdown/summary"
 	"github.com/russross/blackfriday"
 )
 
-const (
-	// DefaultTemplate is the default template.
-	DefaultTemplate = "defaultTemplate"
-	// DefaultStaticDir is the default static directory.
-	DefaultStaticDir = "generated_site"
-)
-
-// Data represents a markdown document.
-type Data struct {
-	middleware.Context
-	Doc      map[string]string
-	DocFlags map[string]bool
-	Links    []PageLink
-}
-
-// Include "overrides" the embedded middleware.Context's Include()
-// method so that included files have access to d's fields.
-func (d Data) Include(filename string) (string, error) {
-	return middleware.ContextInclude(filename, d, d.Root)
+type FileInfo struct {
+	os.FileInfo
+	ctx middleware.Context
 }
 
-// Process processes the contents of a page in b. It parses the metadata
-// (if any) and uses the template (if found).
-func (md Markdown) Process(c *Config, requestPath string, b []byte, ctx middleware.Context) ([]byte, error) {
-	var metadata = newMetadata()
-	var markdown []byte
-	var err error
-
-	// find parser compatible with page contents
-	parser := findParser(b)
-
-	if parser == nil {
-		// if not found, assume whole file is markdown (no front matter)
-		markdown = b
-	} else {
-		// if found, assume metadata present and parse.
-		markdown, err = parser.Parse(b)
-		if err != nil {
-			return nil, err
-		}
-		metadata = parser.Metadata()
+func (f FileInfo) Summarize(wordcount int) (string, error) {
+	fp, err := f.ctx.Root.Open(f.Name())
+	if err != nil {
+		return "", err
 	}
+	defer fp.Close()
 
-	// if template is not specified, check if Default template is set
-	if metadata.Template == "" {
-		if _, ok := c.Templates[DefaultTemplate]; ok {
-			metadata.Template = DefaultTemplate
-		}
+	buf, err := ioutil.ReadAll(fp)
```

Comment:
Having said that, the parts of the code that want a `[]byte`, vs. `string` vs. `io.Reader` vs. `*bytes.Buffer` always end up making this type of code look more ugly then I like.  I really wish that less `[]byte` was used in general...